README.txt
= gitjour

* http://github.com/technomancy/gitjour

== DESCRIPTION:

Automates zeroconf-powered serving and cloning of git repositories.

== SYNOPSIS:

  % gitjour serve [project_dir] [name_to_advertise_as]
  Registered phil-gitjour on port 9418. Starting service.

  % gitjour list
  Gathering for up to 5 seconds...
  === phil-gitjour on pdp10.local.:9418 ===
    gitjour (clone|pull) phil-gitjour

  % gitjour clone phil-gitjour
  Gathering for up to 5 seconds...
  Connecting to pdp10.local.:9418
  Initialized empty Git repository in /home/phil/phil-gitjour/.git/
  [...]
  Resolving deltas: 100% (342/342), done.

  % gitjour pull phil-gitjour
  Gathering for up to 5 seconds...
  Connecting to pdp10.local.:9418
  From git://pdp10.local.:9418
   * branch            master     -> FETCH_HEAD
  Updating 5ede61a..781bc4d
  [...]
  4 files changed, 35 insertions(+), 80 deletions(-)

== REQUIREMENTS:

* git
* dnssd, which requires development headers to bonjour or avahi
  See dnssd documentation for details.

== INSTALL:

* sudo gem install gitjour

== LICENSE:

(The MIT License)

Copyright (c) 2008 Chad Fowler, Evan Phoenix, and Rich Kilmer

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 6.5.0 2009-03-17

* Added web command to launch and announce instaweb
* Added browse command to see all available instawebs

== 6.4.0 2009-03-13

* Add pull command
* Support git 1.6+
* Require dnssd 0.7.1+ to support Avahi

== 6.3.0 2008-08-23

* Simplify project layout, remove newgem dependency
* Add a prefix to projects' served name

[...]

== 0.0.1 2008-05-29

* 1 major enhancement:
  * Initial release


test/test_helper.rb
require 'test/unit'
require 'fileutils'
require File.dirname(__FILE__) + '/../lib/gitjour'


test/test_gitjour.rb
require 'net/telnet'
require File.dirname(__FILE__) + '/test_helper.rb'

class TestGitjour < Test::Unit::TestCase
  def test_thread_friendly
    repo = File.dirname(__FILE__) + '/repo'
    port = 3289
    FileUtils.rm_rf repo
    `mkdir -p #{repo}; cd #{repo}; git init`
    
    thread = Thread.new do
      Gitjour::Application.send(:serve, repo, 'test', port)
    end

    sleep 1
    Net::Telnet::new("Host" => "localhost", "Port" => port)

    thread.kill
    assert_raises(Errno::ECONNREFUSED) do
      Net::Telnet::new("Host" => "localhost", "Port" => port)
    end
  end
end


config/hoe.rb
require 'gitjour/version'

AUTHOR = ['Chad Fowler', 'Rich Kilmer', 'Evan Phoenix']  # can also be an array of Authors
EMAIL = "chad@chadfowler.com"
DESCRIPTION = "Automates DNSSD-powered serving and cloning of git repositories."
GEM_NAME = 'gitjour' # what ppl will type to install your gem
RUBYFORGE_PROJECT = 'gitjour' # The unix name for your project
HOMEPATH = "http://#{RUBYFORGE_PROJECT}.rubyforge.org"
DOWNLOAD_PATH = "http://rubyforge.org/projects/#{RUBYFORGE_PROJECT}"

@config_file = "~/.rubyforge/user-config.yml"
@config = nil
RUBYFORGE_USERNAME = "chadfowler"
def rubyforge_username
  unless @config
    begin
      @config = YAML.load(File.read(File.expand_path(@config_file)))
    rescue
      puts <<-EOS
ERROR: No rubyforge config file found: #{@config_file}
Run 'rubyforge setup' to prepare your env for access to Rubyforge
 - See http://newgem.rubyforge.org/rubyforge.html for more details
      EOS
      exit
    end
  end
  RUBYFORGE_USERNAME.replace @config["username"]
end


REV = nil
# UNCOMMENT IF REQUIRED:
# REV = `svn info`.each {|line| if line =~ /^Revision:/ then k,v = line.split(': '); break v.chomp; else next; end} rescue nil
VERS = Gitjour::VERSION::STRING + (REV ? ".#{REV}" : "")
RDOC_OPTS = ['--quiet', '--title', 'gitjour documentation',
             "--opname", "index.html",
             "--line-numbers",
             "--main", "README",
             "--inline-source"]

class Hoe
  def extra_deps
    @extra_deps.reject! { |x| Array(x).first == 'hoe' }
    @extra_deps
  end
end

# Generate all the Rake tasks
# Run 'rake -T' to see list of generated tasks (from gem root directory)
hoe = Hoe.new(GEM_NAME, VERS) do |p|
  p.author = AUTHOR
  p.description = DESCRIPTION
  p.email = EMAIL
  p.summary = DESCRIPTION
  p.url = HOMEPATH
  p.rubyforge_name = RUBYFORGE_PROJECT if RUBYFORGE_PROJECT
  p.test_globs = ["test/**/test_*.rb"]
  p.clean_globs |= ['**/.*.sw?', '*.gem', '.config', '**/.DS_Store']  #An array of file patterns to delete on clean.
  p.extra_deps =  ['dnssd']
  # == Optional
  p.changes = p.paragraphs_of("History.txt", 0..1).join("\n\n")
  #p.extra_deps = []     # An array of rubygem dependencies [name, version], e.g. [ ['active_support', '>= 1.3.1'] ]

  #p.spec_extras = {}    # A hash of extra values to set in the gemspec.
  p.extra_dev_deps = ['newgem']
end

CHANGES = hoe.paragraphs_of('History.txt', 0..1).join("\\n\\n")
PATH    = (RUBYFORGE_PROJECT == GEM_NAME) ? RUBYFORGE_PROJECT : "#{RUBYFORGE_PROJECT}/#{GEM_NAME}"
hoe.remote_rdoc_dir = File.join(PATH.gsub(/^#{RUBYFORGE_PROJECT}\/?/,''), 'rdoc')
hoe.rsync_args = '-av --delete --ignore-errors'


lib/gitjour/browser.rb
require "webrick"
require "erb"
require "set"
require "thread"

module Gitjour
  class Browser

    def initialize(*args)
      @port = args.shift || 9850
      @browser = args.shift
      @services = Set.new
      @mutex = Mutex.new
    end

    def start
      DNSSD.browse("_http._tcp") do |reply|
        begin
          DNSSD.resolve reply.name, reply.type, reply.domain do |resolve_reply|
            next unless resolve_reply.text_record['gitjour']
            service = GitService.new(reply.name,
                                     resolve_reply.target,
                                     resolve_reply.port,
                                     resolve_reply.text_record['description'].to_s)

            @mutex.synchronize do
              if @services.member? service
                @services.delete service
              else
                @services << service
              end
            end
          end
        rescue ArgumentError # usually a jacked DNS text record
        end
      end

      http = WEBrick::HTTPServer.new(:Port => @port.to_i)
      http.mount_proc("/") { |req, res| index(req, res) }
      http.mount_proc("/style.css") { |req, res| stylesheet(req, res) }
      trap("INT") { http.shutdown }
      t = Thread.new { http.start }

      url = "http://localhost:#{@port}"
      if @browser
        `git web--browse -b '#{@browser}' http://localhost:9850`
      else
        `git web--browse -c "instaweb.browser" http://localhost:9850`
      end
      t.join
    end

    def index(req, res)
      res['Content-Type'] = 'text/html'
      res.body = index_html.result(binding)
    end

    def index_html
      @index_html ||= ERB.new(<<-HTML)
        <html>
          <body>
            <head>
              <link rel="stylesheet" href="/style.css" type="text/css" media="screen"/>
              <title>Browseable Git Repositories</title>
            </head>
            <h1>Browseable Git Repositories</h1>
            <ul>
            <% @mutex.synchronize do %>
              <% @services.map do |s| %>
                <li>
                  <a href='http://<%= s.host %>:<%= s.port %>' target="_new">
                    <%= s.name %>
                  </a>
                  <%= s.description unless s.description =~ /^Unnamed repository/ %>
                </li>
              <% end %>
            <% end %>
            </ul>
          </body>
        </html>
      HTML
    end

    def stylesheet(req, res)
      res['Content-Type'] = 'text/css'
      res.body = css
    end

    def css
      @css ||= <<-CSS
        body {
          font-family: sans-serif;
          font-size: 12px;
          background-color: #fff;
        }

        h1 {
          font-size: 20px;
          font-weight: bold;
        }

        ul {
          border: 1px dashed #999;
          padding: 10 10 10 20;
          background-color: #ccc;
        }
      CSS
    end
  end

end


lib/gitjour/application.rb
require 'rubygems'
require 'dnssd'
require 'set'
require 'webrick'
require 'gitjour/version'

Thread.abort_on_exception = true

module Gitjour
  GitService = Struct.new(:name, :host, :port, :description)  

  class Application

    class << self
      def run(*args)
        case args.shift
          when "list"
            list
          when "clone"
            clone(*args)
          when "serve"
            serve(*args)
          when "remote"
            remote(*args);
          when "web"
            web(*args)
          when "browse"
            browse(*args)
          else
            help
        end
      end

      private
      def list
        service_list("_git._tcp").each do |service|
          puts "=== #{service.name} on #{service.host}:#{service.port} ==="
          puts "  gitjour clone #{service.name}"
          if service.description != '' && service.description !~ /^Unnamed repository/
            puts "  #{service.description}"
          end
          puts
        end
      end

      def clone(repository_name, *rest)
        dir = rest.shift || repository_name
        if File.exists?(dir)
          exit_with! "ERROR: Clone directory '#{dir}' already exists."
        end

        puts "Cloning '#{repository_name}' into directory '#{dir}'..."

        unless service = locate_repo(repository_name)
          exit_with! "ERROR: Unable to find project named '#{repository_name}'"
        end

        puts "Connecting to #{service.host}:#{service.port}"

        system "git clone git://#{service.host}:#{service.port}/ #{dir}/"
      end

      def remote(repository_name, *rest)
        dir = rest.shift || repository_name
        service = locate_repo repository_name
        system "git remote add #{dir} git://#{service.host}:#{service.port}/"
      end

      def serve(path=Dir.pwd, *rest)
        path = File.expand_path(path)
        name = service_name(rest.shift || File.basename(path))
        port = rest.shift || 9418

        if File.exists?("#{path}/.git")
          announce_git(path, name, port.to_i)
        else
          Dir["#{path}/*"].each do |dir|
            if File.directory?(dir)
              name = File.basename(dir)
              announce_git(dir, name, 9418)
            end
          end
        end

        `git-daemon --verbose --export-all --port=#{port} --base-path=#{path} --base-path-relaxed`
      end

      def web(path=Dir.pwd, *rest)
        path = File.expand_path(path)
        name = service_name(rest.shift || File.basename(path))
        port = rest.shift || 1234
        httpd = rest.shift || "webrick"

        if File.exists?("#{path}/.git")
          announce_web(path, name, port.to_i)
          `git-instaweb --httpd=#{httpd} --port=#{port}`
          trap("INT") do 
            puts "Stopping instaweb..."
            `git-instaweb stop`
            exit 0
          end
          while true; sleep 30; end
        else
          $stderr.puts "You must specify a proper git project"
          exit 1
        end
      end

      def service_name(name)
        # If the name starts with ^, then don't apply the prefix
        if name[0] == ?^
          name = name[1..-1]
        else
          prefix = `git config --get gitjour.prefix`.chomp
          prefix = ENV["USER"] if prefix.empty?
          name   = [prefix, name].compact.join("-")
        end
        name
      end

      def help
        puts "Gitjour #{Gitjour::VERSION::STRING}"
        puts "Serve up and use git repositories via Bonjour/DNSSD."
        puts "\nUsage: gitjour <command> [args]"
        puts
        puts "  list"
        puts "      Lists available repositories."
        puts
        puts "  clone <project> [<directory>]"
        puts "      Clone a gitjour served repository."
        puts
        puts "  serve <path_to_project> [<name_of_project>] [<port>] or"
        puts "        <path_to_projects>"
        puts "      Serve up the current directory or projects via gitjour."
        puts
        puts "      The name of your project is automatically prefixed with"
        puts "      `git config --get gitjour.prefix` or your username (preference"
        puts "      in that order). If you don't want a prefix, put a ^ on the front"
        puts "      of the name_of_project (the ^ is removed before announcing)."
        puts
        puts "  web <path_to_project> [<name_of_project>] [<port>] [<httpd_daemon>]"
        puts "      Serve up the current directory via git instaweb for browsers."
        puts "      The default port is 1234 and the httpd_daemon is defaulted to"
        puts "      webrick. Other options are 'lighttpd' and 'apache2' (See the"
        puts "      git-instaweb man page for more details)"
        puts
        puts "  browse [<port>] [<browser>]"
        puts "      Browse git repositories published with the 'web' command (see"
        puts "      above). This command takes two optional arguments: the first"
        puts "      is the port for the local web server (default 9850), the second"
        puts "      is the path to your web browser (see man git-web--browse for"
        puts "      details)."
        puts
        puts "  remote <project> [<name>]"
        puts "      Add a Bonjour remote into your current repository."
        puts "      Optionally pass name to not use pwd."
        puts
      end

      def exit_with!(message)
        STDERR.puts message
        exit!
      end

      class Done < RuntimeError; end

      def discover(type, timeout=5)
        waiting_thread = Thread.current

        dns = DNSSD.browse type do |reply|
          DNSSD.resolve reply.name, reply.type, reply.domain do |resolve_reply|
            service = GitService.new(reply.name,
                                     resolve_reply.target,
                                     resolve_reply.port,
                                     resolve_reply.text_record['description'].to_s)
            begin
              yield service
            rescue Done
              waiting_thread.run
            end
          end
        end

        puts "Gathering for up to #{timeout} seconds..."
        sleep timeout
        dns.stop
      end

      def locate_repo(name)
        found = nil

        discover("_git._tcp") do |obj|
          if obj.name == name
            found = obj
            raise Done
          end
        end

        return found
      end

      def service_list(type)
        list = Set.new
        discover(type) { |obj| list << obj }

        return list
      end

      def browse(*args)
        Browser.new(*args).start
      end

      def http_services
        service_list("_http._tcp,git")
      end

      def announce_git(path, name, port)
        announce_repo(path, name, port, "_git._tcp")
      end

      def announce_web(path, name, port)
        announce_repo(path, name, port, "_http._tcp,git")
      end

      def announce_repo(path, name, port, type)
        return unless File.exists?("#{path}/.git")

        tr = DNSSD::TextRecord.new
        tr['description'] = File.read("#{path}/.git/description") rescue "a git project"

        DNSSD.register(name, type, 'local', port, tr.encode) do |rr|
          puts "Registered #{name} on port #{port}. Starting service."
        end
      end

    end
  end

  class Browser
    require "erb"

    def initialize(*args)
      @port = args.shift || 9850
      @browser = args.shift
      @services = Set.new
      @mutex = Mutex.new
    end

    def start
      DNSSD.browse("_http._tcp,git") do |reply|
        begin
          DNSSD.resolve reply.name, reply.type, reply.domain do |resolve_reply|
            service = GitService.new(reply.name,
                                     resolve_reply.target,
                                     resolve_reply.port,
                                     resolve_reply.text_record['description'].to_s)

            @mutex.synchronize do
              if @services.member? service
                @services.delete service
              else
                @services << service
              end
            end
          end
        rescue ArgumentError # usually a jacked DNS text record
        end
      end

      http = WEBrick::HTTPServer.new(:Port => @port.to_i)
      http.mount_proc("/") { |req, res| index(req, res) }
      http.mount_proc("/style.css") { |req, res| stylesheet(req, res) }
      trap("INT") { http.shutdown }
      t = Thread.new { http.start }

      url = "http://localhost:#{@port}"
      if @browser
        `git web--browse -b '#{@browser}' http://localhost:9850`
      else
        `git web--browse -c "instaweb.browser" http://localhost:9850`
      end
      t.join
    end

    def index(req, res)
      res['Content-Type'] = 'text/html'
      res.body = index_html.result(binding)
    end

    def index_html
      @index_html ||= ERB.new(<<-HTML)
        <html>
          <body>
            <head>
              <link rel="stylesheet" href="/style.css" type="text/css" media="screen"/>
              <title>Browseable Git Repositories</title>
            </head>
            <h1>Browseable Git Repositories</h1>
            <ul>
            <% @mutex.synchronize do %>
              <% @services.map do |s| %>
                <li>
                  <a href='http://<%= s.host %>:<%= s.port %>' target="_new">
                    <%= s.name %>
                  </a>
                  <%= s.description unless s.description =~ /^Unnamed repository/ %>
                </li>
              <% end %>
            <% end %>
            </ul>
          </body>
        </html>
      HTML
    end

    def stylesheet(req, res)
      res['Content-Type'] = 'text/css'
      res.body = css
    end

    def css
      @css ||= <<-CSS
        body {
          font-family: sans-serif;
          font-size: 12px;
          background-color: #fff;
        }

        h1 {
          font-size: 20px;
          font-weight: bold;
        }

        ul {
          border: 1px dashed #999;
          padding: 10 10 10 20;
          background-color: #ccc;
        }
      CSS
    end
  end
end


lib/gitjour.rb
require 'rubygems'
require 'dnssd'
require 'set'

Thread.abort_on_exception = true

module Gitjour
  VERSION = "6.5.0"
  GitService = Struct.new(:name, :host, :port, :description)

  class Application

    class << self
      def run(*args)
        case args.shift
        when "list"
          list
        when "pull"
          pull(*args)
        when "clone"
          clone(*args)
        when "serve"
          serve(*args)
        when "remote"
          remote(*args)
        when "web"
          web(*args)
        when "browse"
          browse(*args)
        else
          help
        end
      end

      private

      def service_name(name)
        # If the name starts with ^, then don't apply the prefix
        if name[0] == ?^
          name = name[1..-1]
        else
          prefix = `git config --get gitjour.prefix`.chomp
          prefix = ENV["USER"] if prefix.empty?
          name   = [prefix, name].compact.join("-")
        end
        name
      end

      def list
        service_list.each do |service|
          puts "=== #{service.name} on #{service.host}:#{service.port} ==="
          puts "  gitjour (clone|pull) #{service.name}"
          if service.description != '' && service.description !~ /^Unnamed repository/
            puts "  #{service.description}"
          end
          puts
        end
      end

      def pull(repository_name, branch = "master")
        service = locate_repo(repository_name) or
          abort "ERROR: Unable to find project named '#{repository_name}'"

        puts "Connecting to #{service.host}:#{service.port}"

        system "git pull git://#{service.host}:#{service.port}/ #{branch}"
      end

      def clone(repository_name, *rest)
        dir = rest.shift || repository_name
        if File.exists?(dir)
          abort "ERROR: Clone directory '#{dir}' already exists."
        end

        puts "Cloning '#{repository_name}' into directory '#{dir}'..."

        service = locate_repo(repository_name) or
          abort "ERROR: Unable to find project named '#{repository_name}'"

        puts "Connecting to #{service.host}:#{service.port}"

        system "git clone git://#{service.host}:#{service.port}/ #{dir}"
      end

      def remote(repository_name, *rest)
        dir = rest.shift || repository_name
        service = locate_repo repository_name
        system "git remote add #{dir} git://#{service.host}:#{service.port}/"
      end

      def serve(path=Dir.pwd, *rest)
        path = File.expand_path(path)
        name = service_name(rest.shift || File.basename(path))
        port = rest.shift || 9418

        if File.exists?("#{path}/.git")
          announce_git(path, name, port.to_i)
        else
          Dir["#{path}/*"].each do |dir|
            if File.directory?(dir)
              name = File.basename(dir)
              announce_git(dir, name, 9418)
            end
          end
        end

        `git daemon --verbose --export-all --port=#{port} --base-path=#{path} --base-path-relaxed`
      end

      def web(path=Dir.pwd, *rest)
        path = File.expand_path(path)
        name = service_name(rest.shift || File.basename(path))
        port = rest.shift || 1234
        httpd = rest.shift || "webrick"

        system("git instaweb --httpd=#{httpd} --port=#{port}") or
          abort "Unable to launch git instaweb."

        announce_web(path, name, port.to_i)

        trap("INT") do
          puts "Stopping instaweb..."
          system "git instaweb stop"
          exit
        end

        Thread.stop
      end

      def help
        puts "Gitjour #{Gitjour::VERSION}"
        puts "Serve up and use git repositories via ZeroConf."
        puts "\nUsage: gitjour <command> [args]"
        puts
        puts "  list"
        puts "      Lists available repositories."
        puts
        puts "  clone <project> [<directory>]"
        puts "      Clone a gitjour-served repository."
        puts
        puts "  serve <path_to_project> [<name_of_project>] [<port>] or"
        puts "        <path_to_projects>"
        puts "      Serve up the current directory or projects via gitjour."
        puts
        puts "      The name of your project is automatically prefixed with"
        puts "      `git config --get gitjour.prefix` or your username (preference"
        puts "      in that order). If you don't want a prefix, put a ^ on the front"
        puts "      of the name_of_project (the ^ is removed before announcing)."
        puts
        puts "  pull <project> [<branch>]"
        puts "      Pull from a gitjour-served repository."
        puts
        puts "  remote <project> [<name>]"
        puts "      Add a ZeroConf remote into your current repository."
        puts "      Optionally pass name to not use pwd."
        puts
        puts "  web <path_to_project> [<name_of_project>] [<port>] [<httpd_daemon>]"
        puts "      Serve up the current directory via git instaweb for browsers."
        puts "      The default port is 1234 and the httpd_daemon is defaulted to"
        puts "      webrick. Other options are 'lighttpd' and 'apache2' (See the"
        puts "      git-instaweb man page for more details)"
        puts
        puts "  browse [<port>] [<browser>]"
        puts "      Browse git repositories published with the 'web' command (see"
        puts "      above). This command takes two optional arguments: the first"
        puts "      is the port for the local web server (default 9850), the second"
        puts "      is the path to your web browser (see man git-web--browse for"
        puts "      details)."
        puts
      end

      class Done < RuntimeError; end

      def discover(timeout=5)
        waiting_thread = Thread.current

        dns = DNSSD.browse "_git._tcp" do |reply|
          DNSSD.resolve reply.name, reply.type, reply.domain do |resolve_reply|
            service = GitService.new(reply.name,
                                     resolve_reply.target,
                                     resolve_reply.port,
                                     resolve_reply.text_record['description'].to_s)
            begin
              yield service
            rescue Done
              waiting_thread.run
            end
          end
        end

        puts "Gathering for up to #{timeout} seconds..."
        sleep timeout
        dns.stop
      end

      def locate_repo(name)
        found = nil

        discover do |obj|
          if obj.name == name
            found = obj
            raise Done
          end
        end

        return found
      end

      def service_list
        list = Set.new
        discover { |obj| list << obj }

        return list
      end

      def browse(*args)
        require "gitjour/browser"
        Browser.new(*args).start
      end

      def announce_repo(path, name, port, type)
        return unless File.exists?("#{path}/.git")

        tr = DNSSD::TextRecord.new
        tr['description'] = File.read("#{path}/.git/description") rescue "a git project"
        tr['gitjour'] = 'true' # distinguish instaweb from other HTTP servers

        DNSSD.register(name, type, 'local', port, tr.encode) do |rr|
          puts "Registered #{name} on port #{port}. Starting service."
        end
      end

      def announce_git(path, name, port)
        announce_repo(path, name, port, "_git._tcp")
      end

      def announce_web(path, name, port)
        announce_repo(path, name, port, "_http._tcp")
      end
    end
  end
end
